util/conditions/setter.go

```diff
@@ -96,6 +96,25 @@ func FalseCondition(t clusterv1.ConditionType, reason string, severity clusterv1
 	}
 }
 
+// NegativeTrueCondition returns a negative polarity condition with Status=True and the given type (Status=True has a negative meaning).
+func NegativeTrueCondition(t clusterv1.ConditionType, reason string, severity clusterv1.ConditionSeverity, messageFormat string, messageArgs ...interface{}) *clusterv1.Condition {
+	return &clusterv1.Condition{
+		Type:     t,
+		Status:   corev1.ConditionTrue,
+		Reason:   reason,
+		Severity: severity,
+		Message:  fmt.Sprintf(messageFormat, messageArgs...),
+	}
+}
+
+// PositiveFalseCondition returns a negative polarity condition with Status=false and the given type (Status=False has a positive meaning).
+func PositiveFalseCondition(t clusterv1.ConditionType) *clusterv1.Condition {
```

Comment:
Why can't this have a reason and message?

Reply:
Back in time we introduced all those helpers to enforce a sort of consistency, and what we have to express when status is good (TrueCondition) does not allow to express reason and message & severity.
This func mirror the same for a condition with negative polarity(when status is False but with a positive meaning).

In other words, I tried to keep things consistent with what we have.

Reply:
Sometimes it can be awkward to require a reason, but there are cases where I think it can make sense.

I guess something like Available as a condition, True status, reason would be AsExpected maybe? But does it also make sense without?

But yeah, in general, I wonder if we should just have the reason and message allowed in all cases, rather than being prescriptive?

Are there kube conventions that say you should or shouldn't set reason/message in certain cases?

Reply:
If a user wants flexibility, he can already do this today using the Set helper.
But as far as I know, in the last 4 years almost everyone used "prescriptive" helpers.

WRT to API conventions, before digging into "reason", let's keep in mind that the scope of this PR is to make one step toward API compliance, not the entire path 😉 

Based on my understanding, reason is another area where the thinking about convention evolved over time. The current field description in corev1 types sort of captures it:

> // reason contains a programmatic identifier indicating the reason for the condition's last transition.
> // Producers of specific condition types may define expected values and meanings for this field,
> // and whether the values are considered a guaranteed API.
> // The value should be a CamelCase string.
> // This field may not be empty.

My reading of this is that community is moving towards having reason everywhere, but they recognize producers of condition might have adopted different approaches.

CAPI as of today has two main differences from the API conventions WRT to how we handle reasons:
- From a semantic point of view, we use reason to "explain" the current state, not the last transition (in some case the diff is subtle, but in controller we just observe what is the current state and surface it in conditions, we do not track how do we get to the current state)
- We set reason only when condition have a negative meaning (as far as I remember setting a reason for positive state was considered redoundant, but it was 4 years ago so I might be wrong...)

This PR intentionally does not mean to change what CAPI today is doing WRT to reasons, because this could imply a way bigger impact that probably we should defer to when we cut a new API version.

Reply:
> This PR intentionally does not mean to change what CAPI today is doing WRT to reasons, because this could imply a way bigger impact that probably we should defer to when we cut a new API version.

This feels sensible for keeping the scope small here 😄 

For my 0.02c:  this approach 'feels' more right to me (less surprising)

> From a semantic point of view, we use reason to "explain" the current state, not the last transition (in some case the diff is subtle, but in controller we just observe what is the current state and surface it in conditions, we do not track how do we get to the current state)